### ISSUE TITLE
Fl_Help_VIew: allow Fl_Scrollbars to be accessable

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#146]
As noted in a post today, it's useful for applications/subclasses to access Fl_Help_View's scrollbars to, for instance, hide() them.

They are currently private: and are instances named scrollbar_ and hscrollbar_.

Options I can think of:

1. Make scrollbars "protected:" or "public:" and named appropriately for direct access.
2. Expose scrollbars via accessor methods that return pointers
3. Expose scrollbars via accessor methods that return references
4. Other?

Precedents:

- Fl_Scroll exposes scrollbar and hscrollbar as public instances
- Fl_Browser exposes "scrollbar" and "hscrollbar" as public instances
- Fl_Text_Display exposes *mHScrollBar and *mVScrollBar as protected pointers
- Fl_Tree exposes *_vscroll and *_hscroll as protected pointers
- Fl_Table exposes *vscrollbar and *hscrollbar as protected pointers

I'd like to hold back my vote on which I prefer for last, just to see what everyone thinks, as I'm not really sure what's best.

What I can say is:
- I don't like camel case exposed in Fl_Text_Display for its scrollbar member names, so let's not do that
- I don't think it's good to expose members with underbars as a prefix or suffix (e.g. Fl_Tree)
- I like the names "scrollbar" and "hscrollbar" for vert/horiz scrollbars respectively. Whether we should expose them as methods or direct access, not sure. I generally prefer using pointers for everything, just so construction time can be specific, but there may be reasons why FL_Help_View uses instances instead of pointers and 'new'.

**Comment by Albrecht-S** [comment]
I vote for `public` accessor methods `scrollbar()` and `hscrollbar()` that return pointers. For "symmetry" reasons I also like `vscrollbar()` and `hscrollbar()` but that's IMHO less consistent with existing widgets.

Reasoning:

1. Exposing member instances directly is (and was) IMHO a bad choice. I'm thinking of a future enhancement so users can replace the internal (default) `Fl_Scrollbar`'s with their own derived classes for whatever reasons (draw methods etc.). This is not possible with the current code using (and exposing) internal member variables. One way to achieve this would be to add accessor methods as mentioned above and to deprecate direct access to the (currently) `public` members. This could be done in FLTK 1.4 and in a later version we could (a) make the member variables `private` and (b) replace them with pointers and allocate them with `new`. In step (c) we could add setter methods so users can actually replace the scrollbars with their own derived classes of scrollbars.

_Note_: It's not possible to use an accessor method, for instance `scrollbar()`, with the same name as the internal member variable `scrollbar`, hence if we needed both for backwards compatibility (for some time) we'd have to use other names, e.g. `pscrollbar()` for the accessor methods which is unfortunately _"not the FLTK way"_. FYI: Compiler (gcc) error message is
```
../../FL/Fl_Scroll.H:141:51: error: ‘Fl_Scrollbar* Fl_Scroll::scrollbar()’ conflicts with a previous declaration
  141 |   Fl_Scrollbar *scrollbar()  { return &scrollbar; }
      |                                                   ^
../../FL/Fl_Scroll.H:138:16: note: previous declaration ‘Fl_Scrollbar Fl_Scroll::scrollbar’
  138 |   Fl_Scrollbar scrollbar;
      |                ^~~~~~~~~
```

2. Since the member variables in Fl_Help_View are already `private` (and named correctly with trailing `'_'`) we can safely add the `public` (or `protected`) accessor methods with appropriate names (as suggested).

3. If we wanted to take the route described in (1) we should, however, use another name as described in (1). To use all possible options (compatibility, consistency, and symmetry) we might want to use `pvscrollbar()` and `phscrollbar()`.

**Comment by Albrecht-S** [comment]
After thinking more about this issue I believe that using "other" method names like `pvscrollbar()` and `phscrollbar()` (better proposals, anybody?) would be the best choice because we can (and should) add these methods to all widgets using internal scrollbars w/o name conflicts.

**Comment by Albrecht-S** [comment]
Coming back to the original issue:
> it's useful for applications/subclasses to access Fl_Help_View's scrollbars to, for instance, hide() them

While we're at it we could **additionally** create methods to switch scrollbars on (`show()`) and off (`hide()`) to have an official way to do this with all proper consequences and not the "hack" to `hide()` them from user code. This doesn't mean that we should not also expose them by accessor methods.

**Comment by erco77** [comment]
`(better proposals, anybody?`

1. v_scrollbar(), h_scrollbar()?
2. scrollbar_v(), scrollbar_h()?

The nice thing about #2's naming is the scrollbar access functions all sort together in the docs alphabetically.

**Comment by erco77** [comment]
(The 'p' prefix on the method names seemed a bit cryptic to me, not immediately evident what it stands for)

I suppose another option to the above (1) and (2) suggs might be:

3.  scrollbar_vert() and scrollbar_horiz()

..though I prefer shorter names when possible.

**Comment by Albrecht-S** [comment]
@erco77: I like your proposals, particularly
> 2. scrollbar_v(), scrollbar_h()

The only thing that comes to mind is that `_h()` might be misinterpreted as "height" rather than "horizontal". I'd be fine with it anyway, but there's another option:

3. scrollbar_x(), scrollbar_y() which stands for "scrollbar in { x, y } direction", resp.

> ... even though exposing the members is perhaps bad design, it's "the fltk way", and perhaps it's best to just accept the warts for what they are, and do a "cleanup" in a rewrite, which is what FLTK 2 I think tried to do (among other things).

This may be considered OT here, but if we wanted to wait for a rewrite to improve things that would probably be forever. :-( We failed with FLTK 2 and FLTK 3 because it's just not doable with all compatibility issues, hence I'd rather try to move forward step by step. That said, ...

I strongly believe that we should add new accessor methods for the scrollbars, but in my opinion they **really should be protected** (after thinking more about it) because fiddling with the scrollbars in user code (if public) can have strange side effects (e.g. the user hides the scrollbars but the widget (class) calculates space although the scrollbars are hidden, or the widget re-enables them during draw() etc. etc.).

But we can (and probably should) also add public methods to disable/enable scrollbars. This can be done safely because the class implementation can still control the scrollbars depending on the enabled/disabled status. W/o looking at the code of other widgets ISTR that we have this in one or more widgets with scrollbars.

**Comment by erco77** [comment]
Good point about the 'height' vs 'horizontal', though I think we're safe as we currently have only these methods that start with scrollbar and end in h():

```
$ git grep scrollbar | grep 'h('
Fl_Browser_.H:  int scrollbar_width() const {
Fl_Browser_.H:  void scrollbar_width(int width) {
Fl_Text_Display.H:  int scrollbar_width() const {
Fl_Text_Display.H:  void scrollbar_width(int width) {
```
But those methods are all deprecated in favor of scrollbar_size(), to universally apply to the "short dimension" for both horizontal and vertical scrollbars. So 'height' and 'width' terminology is generally avoided for scrollbars. (Which is why we don't have scrollbar_height() anywhere in the API)

So I think we might be OK with scrollbar_v()/_h(), or for the longer scrollbar_vert()/_horiz().

> scrollbar_x(), scrollbar_y() which stands for "scrollbar in { x, y } direction", resp.

I'm thinking that'd be even more easily confused with the x() and y() positions, so -1 on that.

**Comment by Albrecht-S** [comment]
> Good point about the 'height' vs 'horizontal', though I think we're safe as we currently have only these methods that start with scrollbar and end in h(): [...]

My concern is not of technical nature, I'm concerned about users being confused by `h()` which is perpendicular to `scrollbar_h()` but that's only a question of the best choice and good documentation.

**Comment by MatthiasWM** [comment]
New methods are Fl_Help_View::scrollbar() and Fl_Help_View::hscrollbar(), taking inspiration from Fl_Browser.
